File: worker/satellite.py
"""Analyse 'potentiel solaire' d'un lead via vue aérienne IGN + Claude Vision.

Utilisé par le worker en mode réel quand WORKER_SATELLITE est activé. Volontairement
tolérant : toute erreur renvoie None (l'enrichissement est optionnel, jamais bloquant).
"""

from __future__ import annotations

import base64
import json
import logging
import math
import os
from datetime import datetime, timezone
from typing import Any
from urllib.parse import urlencode

import requests

log = logging.getLogger("renoboost.worker.satellite")

_WMS = "https://data.geopf.fr/wms-r/wms"
_R = 6378137.0
_DEMI_COTE_M = 130.0
_MODEL = os.environ.get("WORKER_SATELLITE_MODEL", "claude-haiku-4-5")

_PROMPT = (
    "Tu analyses une vue aérienne IGN (orthophoto) centrée sur le site d'une entreprise, "
    "pour Rossini Energy (solaire en toiture, ombrières de parking, bornes de recharge). "
    "Décris UNIQUEMENT ce que tu OBSERVES (n'invente rien, ne calcule pas de score).\n"
    "Réponds UNIQUEMENT par un objet JSON valide, sans texte autour, au format :\n"
    '{"toiture":{"presente":<bool>,'
    '"type":"plate|inclinee|inclinee_nord|encombree|inconnue",'
    '"surface_estimee_m2":<number|null>,"ratio_exploitable":<number 0-1|null>},'
    '"parking":{"present":<bool>,"surface_estimee_m2":<number|null>,'
    '"nb_places_estime":<integer|null>,"partage":<bool|null>},'
    '"bornes_visibles":<integer>}\n'
    "ratio_exploitable = part de toiture réellement équipable (hors édicules, "
    "lanterneaux, ombres). partage = parking partagé entre plusieurs enseignes/occupants. "
    "bornes_visibles = nombre de bornes de recharge VE visibles sur le site. "
    "Mets null si tu n'es pas sûr d'une valeur chiffrée."
)


def _ign_url(lat: float, lon: float) -> str:
    x = _R * math.radians(lon)
    y = _R * math.log(math.tan(math.pi / 4 + math.radians(lat) / 2))
    bbox = f"{x - _DEMI_COTE_M},{y - _DEMI_COTE_M},{x + _DEMI_COTE_M},{y + _DEMI_COTE_M}"
    params = {
        "SERVICE": "WMS",
        "VERSION": "1.3.0",
        "REQUEST": "GetMap",
        "LAYERS": "ORTHOIMAGERY.ORTHOPHOTOS",
        "STYLES": "",
        "CRS": "EPSG:3857",
        "BBOX": bbox,
        "WIDTH": "640",
        "HEIGHT": "640",
        "FORMAT": "image/jpeg",
    }
    return f"{_WMS}?{urlencode(params)}"


def _parse_json(text: str) -> dict[str, Any] | None:
    try:
        return json.loads(text)
    except Exception:  # noqa: BLE001
        debut = text.find("{")
        fin = text.rfind("}")
        if debut != -1 and fin > debut:
            try:
                return json.loads(text[debut : fin + 1])
            except Exception:  # noqa: BLE001
                return None
        return None


def _vision_observations(
    lat: float, lon: float, api_key: str, timeout: float
) -> tuple[dict[str, Any], str] | None:
    """Appelle Claude Vision sur l'orthophoto → (observations, image_url)."""
    url = _ign_url(lat, lon)
    img = requests.get(url, timeout=timeout)
    if img.status_code != 200 or len(img.content) < 1000:
        return None
    b64 = base64.b64encode(img.content).decode("ascii")
    res = requests.post(
        "https://api.anthropic.com/v1/messages",
        headers={
            "x-api-key": api_key,
            "anthropic-version": "2023-06-01",
            "content-type": "application/json",
        },
        json={
            "model": _MODEL,
            "max_tokens": 700,
            "messages": [
                {
                    "role": "user",
                    "content": [
                        {
                            "type": "image",
                            "source": {
                                "type": "base64",
                                "media_type": "image/jpeg",
                                "data": b64,
                            },
                        },
                        {"type": "text", "text": _PROMPT},
                    ],
                }
            ],
        },
        timeout=timeout,
    )
    if res.status_code != 200:
        log.warning("Vision satellite HTTP %s", res.status_code)
        return None
    data = res.json()
    text = "".join(
        b.get("text", "") for b in data.get("content", []) if b.get("type") == "text"
    )
    parsed = _parse_json(text)
    return (parsed, url) if parsed is not None else None


def analyser_potentiel(
    lat: float,
    lon: float,
    api_key: str,
    timeout: float = 30.0,
    *,
    signaux_ve: bool = False,
    exploitant: str | None = None,
) -> dict[str, Any] | None:
    """Analyse les 3 potentiels (solaire/ombrières/bornes) → dict, ou None.

    Combine l'observation Claude Vision (orthophoto IGN), la surface d'emprise
    bâtie IGN et le comptage de bornes IRVE, puis applique le scoring /10.
    Tolérant : toute panne renvoie None et n'interrompt jamais le run.
    """
    from renoboost_leads.potentiel.analyse import analyser_potentiels
    from renoboost_leads.potentiel.ign import surface_batie_m2
    from renoboost_leads.potentiel.irve import ClientIRVE
    from renoboost_leads.potentiel.solar_api import surfaces_solar_api

    try:
        obs = _vision_observations(lat, lon, api_key, timeout)
        if obs is None:
            return None
        vision, url = obs
        # Surface toiture : Google Solar API si dispo (précis), sinon emprise IGN.
        solar = surfaces_solar_api(lat, lon)
        if solar is not None:
            surface_toiture, surface_exploitable = solar
        else:
            surface_toiture, surface_exploitable = surface_batie_m2(lat, lon), None
        analyse = analyser_potentiels(
            latitude=lat,
            longitude=lon,
            vision=vision,
            signaux_ve=signaux_ve,
            irve=ClientIRVE(),
            surface_batie_m2=surface_toiture,
            surface_exploitable_m2=surface_exploitable,
            exploitant_probable=exploitant,
            image_url=url,
            analyzed_at=datetime.now(timezone.utc).isoformat(),
        )
        return analyse.model_dump()
    except Exception:  # noqa: BLE001 — enrichissement optionnel, jamais bloquant
        log.exception("Analyse satellite échouée")
        return None


def satellite_active() -> bool:
    return os.environ.get("WORKER_SATELLITE", "").strip().lower() in ("1", "true", "yes")


def satellite_max() -> int:
    try:
        return int(os.environ.get("WORKER_SATELLITE_MAX", "60"))
    except ValueError:
        return 60

File: web/lib/satellite.ts
import type { SupabaseClient } from "@supabase/supabase-js";

// Analyse "potentiel solaire" d'un lead via vue aérienne IGN + Claude Vision.
// Partagé entre la route /api/lead/satellite et l'outil Magellan.

const MODEL = "claude-haiku-4-5";
const R = 6378137; // rayon terrestre (Web Mercator)
const DEMI_COTE_M = 130; // demi-côté de la vue (~260 m de large)

function toMercator(lon: number, lat: number): [number, number] {
  const x = (R * (lon * Math.PI)) / 180;
  const y = R * Math.log(Math.tan(Math.PI / 4 + (lat * Math.PI) / 360));
  return [x, y];
}

export function ignUrl(lat: number, lon: number): string {
  const [x, y] = toMercator(lon, lat);
  const bbox = [x - DEMI_COTE_M, y - DEMI_COTE_M, x + DEMI_COTE_M, y + DEMI_COTE_M].join(",");
  const params = new URLSearchParams({
    SERVICE: "WMS",
    VERSION: "1.3.0",
    REQUEST: "GetMap",
    LAYERS: "ORTHOIMAGERY.ORTHOPHOTOS",
    STYLES: "",
    CRS: "EPSG:3857",
    BBOX: bbox,
    WIDTH: "640",
    HEIGHT: "640",
    FORMAT: "image/jpeg",
  });
  return `https://data.geopf.fr/wms-r/wms?${params.toString()}`;
}

const PROMPT = `Tu analyses une vue aérienne IGN (orthophoto) centrée sur le site d'une entreprise, pour évaluer son potentiel solaire (panneaux en toiture et ombrières de parking).
Réponds UNIQUEMENT par un objet JSON valide, sans texte autour, au format :
{"score":<0-100>,"verdict":"<phrase courte>","toiture":{"presente":<bool>,"type":"plate|inclinee|inconnue","surface_estimee_m2":<number|null>},"parking":{"present":<bool>,"surface_estimee_m2":<number|null>,"ombrieres_possibles":<bool>},"commentaire":"<2-3 phrases>"}
Le score reflète l'intérêt global (grandes surfaces planes = élevé). Sois prudent si l'image est ambiguë.`;

function parseJson(text: string): Record<string, unknown> | null {
  try {
    return JSON.parse(text);
  } catch {
    const m = text.match(/\{[\s\S]*\}/);
    if (m) {
      try {
        return JSON.parse(m[0]);
      } catch {
        return null;
      }
    }
    return null;
  }
}

type AnalyseResult =
  | { error: string }
  | { ok: true; result: Record<string, unknown> };

export async function analyseSatellite(
  supabase: SupabaseClient,
  leadId: string,
): Promise<AnalyseResult> {
  const apiKey = process.env.ANTHROPIC_API_KEY;
  if (!apiKey)
    return { error: "Analyse indisponible : ANTHROPIC_API_KEY manquante côté serveur." };

  const { data: lead } = await supabase
    .from("leads")
    .select("id, entreprise, adresse, ville, code_postal, latitude, longitude")
    .eq("id", leadId)
    .single();
  if (!lead) return { error: "Lead introuvable." };

  // 1) Coordonnées : connues, sinon géocodage BAN.
  let lat = lead.latitude as number | null;
  let lon = lead.longitude as number | null;
  if (lat == null || lon == null) {
    const q = [lead.adresse || lead.entreprise, lead.ville].filter(Boolean).join(" ");
    if (!q.trim()) return { error: "Pas d'adresse pour localiser ce lead." };
    const url = new URL("https://api-adresse.data.gouv.fr/search/");
    url.searchParams.set("q", q);
    url.searchParams.set("limit", "1");
    if (lead.code_postal) url.searchParams.set("postcode", String(lead.code_postal));
    try {
      const geo = await fetch(url.toString());
      const gj = await geo.json();
      const coords = gj?.features?.[0]?.geometry?.coordinates;
      if (Array.isArray(coords)) {
        lon = coords[0];
        lat = coords[1];
      }
    } catch {
      /* ignore */
    }
    if (lat == null || lon == null)
      return { error: "Localisation impossible (géocodage échoué)." };
    await supabase.from("leads").update({ latitude: lat, longitude: lon }).eq("id", leadId);
  }

  const imageUrl = ignUrl(lat, lon);

  // 2) Image IGN → base64.
  let b64: string;
  try {
    const img = await fetch(imageUrl);
    if (!img.ok) throw new Error(`IGN ${img.status}`);
    const buf = Buffer.from(await img.arrayBuffer());
    if (buf.length < 1000) throw new Error("image vide");
    b64 = buf.toString("base64");
  } catch (e) {
    console.error("IGN fetch", e);
    return { error: "Image satellite indisponible pour ce point." };
  }

  // 3) Claude Vision.
  try {
    const res = await fetch("https://api.anthropic.com/v1/messages", {
      method: "POST",
      headers: {
        "content-type": "application/json",
        "x-api-key": apiKey,
        "anthropic-version": "2023-06-01",
      },
      body: JSON.stringify({
        model: MODEL,
        max_tokens: 700,
        messages: [
          {
            role: "user",
            content: [
              {
                type: "image",
                source: { type: "base64", media_type: "image/jpeg", data: b64 },
              },
              { type: "text", text: PROMPT },
            ],
          },
        ],
      }),
    });
    if (!res.ok) {
      console.error("Anthropic vision", res.status, await res.text());
      return { error: "Analyse IA momentanément indisponible." };
    }
    const data = await res.json();
    const text = (data.content ?? [])
      .filter((b: { type: string }) => b.type === "text")
      .map((b: { text: string }) => b.text)
      .join("\n");
    const parsed = parseJson(text);
    if (!parsed) return { error: "Réponse IA illisible, réessaie." };

    const result = { ...parsed, image_url: imageUrl, analyse_le: new Date().toISOString() };
    await supabase.from("leads").update({ vision_satellite: result }).eq("id", leadId);
    return { ok: true, result };
  } catch (e) {
    console.error("vision error", e);
    return { error: "Erreur lors de l'analyse." };
  }
}
